### PR TITLE
ARROW-848: [Python] Another pass on conda dev guide

### PR DESCRIPTION
**Pull request by wesm**
Per feedback in https://github.com/apache/arrow/commit/bb8514cc9d7068c8b62d346577370751d68221d8

### CODE REVIEW

File: python/DEVELOPMENT.md [hunk 1]
```diff
@@ -67,15 +45,36 @@ drwxrwxr-x 12 wesm wesm 4096 Apr 15 19:19 arrow/
 drwxrwxr-x 12 wesm wesm 4096 Apr 15 19:19 parquet-cpp/
 ```
 
+We need to set a number of environment variables to let Arrow's build system
+know about our build toolchain:
+
+```
+export ARROW_BUILD_TYPE=release
+
+export BOOST_ROOT=$CONDA_PREFIX
```
Comment (xhochy) [inline python/DEVELOPMENT.md]: We definitely should invest into getting all those variables into a single one. Better would be even that we could detect if we're in a conda env and try to prefer the libraries installed there.

Comment (xhochy) [inline python/DEVELOPMENT.md]: We could e.g. just try and search first for all dependencies in `$CMAKE_INSTALL_PREFIX`

Comment (wesm) [inline python/DEVELOPMENT.md]: What if we had an "ARROW_BUILD_TOOLCHAIN" that automatically set all of the *_HOME variables?

Comment (xhochy) [inline python/DEVELOPMENT.md]: Yes, that is in scope I would hope for :)

Comment (wesm) [inline python/DEVELOPMENT.md]: see ARROW-849 and PARQUET-957

**Comment by xhochy** [comment]
I can reproduce @mrocklin problem on a fresh Ubuntu 17.04 installation: https://gist.github.com/xhochy/7bb0e4d47cbe25b6d79a3d6b1db7a9ea

**Comment by wesm** [comment]
Using these latest instructions?

**Comment by xhochy** [comment]
Yes, made a JIRA about this as I probably don't have time today to fix it: https://issues.apache.org/jira/browse/ARROW-850

**Comment by wesm** [comment]
If you build with `ARROW_BOOST_USE_SHARED=on` do things work in this environment? I can set up a VM later to try it out myself

**Comment by xhochy** [comment]
This leads to another problem (C++ ABI mismatch): https://gist.github.com/xhochy/01d0a8cbbea3f38cbb85dbb70a5ec2a5

**Comment by wesm** [comment]
Right, you can't use gcc >= 5 with the conda-forge stack (unless you disable the gcc5 ABI), so we'll need to include instructions to either install gcc 4.9 on newer Linuxes or add compiler flags.

**Comment by xhochy** [comment]
`cmake -DARROW_BOOST_USE_SHARED=on CMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=0"` did the trick.

**Comment by xhochy** [comment]
I was using gcc-5 before but on an older Ubuntu version that by default was using the old ABI. The new Ubuntu release switched the default ABI.

**Comment by wesm** [comment]
excellent intel, thanks @xhochy!

**Comment by xhochy** [comment]
In my setup, this still needs `export LD_LIBRARY_PATH=$CONDA_PREFIX/lib`

**Comment by xhochy** [comment]
JIRA for the CXX ABI: https://issues.apache.org/jira/browse/ARROW-851

**Comment by wesm** [comment]
@xhochy at what point do you need the LD_LIBRARY_PATH?

**Comment by xhochy** [comment]
For running the tests in pyarrow.

**Comment by wesm** [comment]
Very odd. And the conda environment is activated?

**Comment by xhochy** [comment]
Yes

**Comment by wesm** [comment]
@xhochy I figured out the problem -- after the libarrow_python refactor, for conda-based setups, **we do not need to modify the Cython module RPATH**. I think you still need to do this for manylinux1 builds. I found that removing

```
    if(APPLE)
        set(module_install_rpath "@loader_path")
    else()
        set(module_install_rpath "\$ORIGIN")
    endif()
    list(LENGTH directories i)
    while(${i} GREATER 0)
        set(module_install_rpath "${module_install_rpath}/..")
        math(EXPR i "${i} - 1" )
    endwhile(${i} GREATER 0)

    set_target_properties(${module_name} PROPERTIES
      INSTALL_RPATH ${module_install_rpath})
```

solved the problem. I'm opening a new JIRA about this

File: python/DEVELOPMENT.md [hunk 2]
```diff
@@ -16,36 +16,41 @@
...
-export BROTLI_HOME=$CPP_TOOLCHAIN
-export JEMALLOC_HOME=$CPP_TOOLCHAIN
-export ARROW_HOME=$CPP_TOOLCHAIN
-export PARQUET_HOME=$CPP_TOOLCHAIN
+Finally, set gcc 4.9 as the active compiler using:
 
-conda create -y -q -p $CPP_TOOLCHAIN \
-      flatbuffers rapidjson boost-cpp thrift-cpp snappy zlib brotli jemalloc
+```shell
+export CC=gcc-4.9
```
Comment (cpcloud) [inline python/DEVELOPMENT.md]: Should this be `export CC=$(which gcc-4.9)`? (and same for `CXX`)

File: python/DEVELOPMENT.md [hunk 3]
```diff
@@ -67,14 +72,34 @@ drwxrwxr-x 12 wesm wesm 4096 Apr 15 19:19 arrow/
...
 
 ```shell
 mkdir arrow/cpp/build
 pushd arrow/cpp/build
 
 cmake -DCMAKE_BUILD_TYPE=$ARROW_BUILD_TYPE \
-      -DCMAKE_INSTALL_PREFIX=$CPP_TOOLCHAIN \
+      -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX \
       -DARROW_PYTHON=on \
       -DARROW_BUILD_TESTS=OFF \
```
Comment (cpcloud) [inline python/DEVELOPMENT.md]: Tiny nit: Maybe make the case consistent here.

File: python/DEVELOPMENT.md [hunk 4]
```diff
@@ -67,14 +72,34 @@ drwxrwxr-x 12 wesm wesm 4096 Apr 15 19:19 arrow/
...
 ```
 
+We need to set a number of environment variables to let Arrow's build system
+know about our build toolchain:
+
+```
+export ARROW_BUILD_TYPE=release
+
+export BOOST_ROOT=$CONDA_PREFIX
+export BOOST_LIBRARYDIR=$CONDA_PREFIX/lib
```
Comment (cpcloud) [inline python/DEVELOPMENT.md]: Will this need to be `lib64` when the system is 64 bit?

Comment (wesm) [inline python/DEVELOPMENT.md]: conda-forge hardcodes lib https://github.com/conda-forge/boost-cpp-feedstock/blob/master/recipe/build.sh#L14

File: python/DEVELOPMENT.md (hunk 2 above)
Comment (wesm) [inline python/DEVELOPMENT.md]: either work, but I'll change to the more explicit variety.

Comment (wesm) [inline python/DEVELOPMENT.md]: The output of $(which FOO) doesn't produce an executable command for me when FOO is found multiple times in the path. I'll leave this as is for now

```
$ which gcc
gcc is /home/wesm/.conda/envs/$NAME/bin/gcc
gcc is /usr/bin/gcc
```

File: python/DEVELOPMENT.md (hunk 4 above)
Comment (cpcloud) [inline python/DEVELOPMENT.md]: Cool.

File: python/DEVELOPMENT.md (hunk 2 above)
Comment (cpcloud) [inline python/DEVELOPMENT.md]: Sounds good.

**Comment by wesm** [comment]
+1. We can improve after adding the *_TOOLCHAIN variables

**Pull request closed without merging**